core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java

```diff
@@ -159,16 +162,79 @@ private List<Result> createScanResults(Scan scan, List<String> projections, List
         .collect(Collectors.toList());
   }
 
-  public void put(Put put) throws UnsatisfiedConditionException {
-    mutationConditionsValidator.checkIfConditionIsSatisfied(
-        put, snapshot.getFromReadSet(new Snapshot.Key(put)).orElse(null));
-    snapshot.put(new Snapshot.Key(put), put);
+  public void put(Put put) throws CrudException {
+    if (put.getCondition().isPresent() && put.isBlind()) {
+      throw new IllegalArgumentException("Blind put cannot have a condition: " + put);
+    }
+
+    Snapshot.Key key = new Snapshot.Key(put);
+
+    if (put.getCondition().isPresent()) {
+      fillReadSetIfUnread(key);
+      mutationConditionsValidator.checkIfConditionIsSatisfied(
+          put, snapshot.getFromReadSet(key).orElse(null));
+    }
+
+    snapshot.put(key, put);
+  }
+
+  public void delete(Delete delete) throws CrudException {
+    Snapshot.Key key = new Snapshot.Key(delete);
+
+    if (delete.getCondition().isPresent()) {
+      fillReadSetIfUnread(key);
```

Comment:
QQ: Why only if a condition is set?
This might be a silly idea or I might be missing something, but how about always calling `fillReadSetIfUnread()` in `delete()` and `put()` even if no condition is set so that a commit doesn't need to call `fillReadSetForRecordsFromWriteAndDeleteSetsIfUnread()`?

Reply:
Indeed, we can fill the read set anytime before preparing records. So, as you suggested, we could always call `fillReadSetIfUnread()` in `delete()` and `put()`. The benefit of calling `fillReadSetForRecordsFromWriteAndDeleteSetsIfUnread()` before preparing records is that we can fill the read set in parallel for better performance. So we basically fill the read set right before preparing records, but for conditional updates, we need to fill the read set at that time.

Reply:
Understood. Sounds reasonable 👍